### DISCUSSION TITLE
OpenPlotter integration

### DISCUSSION THREAD

**Discussion by peterantypas** [peterantypas/maiana#81] [General]
I am happy to announce that I have been in direct discussions with Roberto (aka Sailoog) from the OpenPlotter project, and we have ironed out a path to make MAIANA a built-in AIS solution for OP.

The main idea is that MAIANA will be managed entirely by a graphical interface in OpenPlotter, and the hardware interface will be a really inexpensive breakout board with direct UART connection to the Raspberry Pi.

In addition, we'll be exploring the possibility of distributing the MAIANA kit to European customers directly from the OpenMarine store, so as to avoid high shipping fees from California. 

Personally, I think it's an ideal match and I'm very excited to get started!

**Comment by marcelrv** [comment]
That is great news. Honestly the price incl very highly shipping cost to EU were at least to me prohibitive

**Comment by sailoog** [comment]
Yeah, MAIANA should be a "must have" for OpenPlotter if we want to achieve the full open source boat and we are close to that, we just need to find a solid solution for the sounder.

This new OpenPlotter MAIANA app will take control of the serial connection of the USB/0183/2000/UART adapters before feeding the Signal K server and it will make possible enable/disable transmission and program the transponder with the boat data from the Signal K server or from a form. You will also be able to display transponder data and status on sexy dashboards or set virtual switches to enable/disable the transponder.

Once this data is on the Signal K server you could also make MAIANA interact with the rest of the data of your boat and set triggers and actions. The possibilities are endless, for example you could: establish geofences and only transmit in some areas, only transmit when you have a close target, only transmit at night, disable transmission when you detect a ship known as "hostile"... 

it will depend on the dimensions of the packages but I am pretty sure we can send MAIANAs from Barcelona to any EU country for €5-7 for ordinary tracked shipments (5-10 days) and €19-25 for express shipments (2 days).

Comments and suggestions are welcome.

**Reply by muebau** [reply]
Good to hear this. As a "early adopter" I get sad if I hear the prices, especially as I decided to use a NMEA2000 version :-) So it sounds like the existing devices will be compatible with the new system, right? I simply USB connection with OpenPlotter.

BTW: Yes a open source sounder and open source log (speed through water) is the last thing missing. I need to replace both in the future and cant find anything good.
I thing the depth meter would be a simple thing if build like the old NASA Sting Ray devices. A rotating light will fire an impulse when at "12 position" and the echo lights the light on the rotating arm. If done by a uC it should be relatively easy generate the impulse and record the echos with ADC. This way every responce could be shown as a stripe and the compete picture would be shown in OpenPlotter as a "depth sound picture" over time.

**Reply by sailoog** [reply]
Clarification: The prices I am talking about are only for shipments between EU, not for the transponder and adapters.

Yes, existing devices will be compatible, they will only require a firmware update. Peter will add more information on this.

I have already seen some open source project for sounder / log on small hobbyist boats, we just need a project like these that will scale to large boats.

**Reply by muebau** [reply]
Oh OK.

I am happy to hear the compatibility is easy.

If you have any hints for  sounder / log projects I would love to get anything. I googled a lot for month and couldn't even find simple circuits to generate an impulse for my transducer safely or find a way to reuse the existing log hardware build in the ship.

**Reply by Techstyleuk** [reply]
Great news!

Will the integration mean that Maiana's internal GPS can be use and remove the need for an external GPS source?

I am sure you are aware, but just incase - The Marine Traffic plugin already manages creating the outgoing messages needed for AIS transmission in SignalK, this could be used a a base if you haven't already done this

please keep us up to date with progress and availability - I am an eager tester.

Thanks

**Comment by peterantypas** [comment]
The main update for MAIANA to support OpenPlotter is firmware with additional commands, so all existing breakouts could be wired directly to a RPi via USB and work over serial.

Now, for OpenPlotter specifically, we're working on a very simple and inexpensive breakout board that just exposes the raw signals and UART, so it will reduce the overall kit cost somewhat. We still have to figure out the logistics of distributing the kit from Europe.

**Comment by peterantypas** [comment]
Regarding sounder, I'm also interested. Just started the retrofit process last week and I'm planning a haulout in the next two months, during which I want to replace all my really old instruments. In the past, I was semi-successful prototyping a circuit for my old Datamarine 200KHz transducer which measured to about 60ft (it was on a breadboard). But I don't really have bandwidth to take on such a project right now.

**Reply by muebau** [reply]
Do you have any circuit diagrams? I was not brave enough to play with potential very high voltage peaks of the transducer. I would like to leave the old transducer in the ship but replace the electronics and give it a new live this way.

**Reply by peterantypas** [reply]
I'm afraid not, never really moved to a formal design.

**Comment by muebau** [comment]
Just an idea:
Is there a plan to make it possible to push firmware updates to MAIANA with OpenPlotter. Like Linux does it with "linux-firmware" there could be a package to bring updates to the firmware of the MAIANA hardware. There could be some command to reboot the uC to a firmware bootloader (like done with the reset button currently) and write the firmware with the same protocol as the STM software does it.

What do you think?

**Reply by sailoog** [reply]
According to this, that could be possible: https://github.com/peterantypas/maiana/blob/master/latest/Firmware/Transponder/Python/bootloader.py

I have not any transponder and I can not test this yet. Peter will have more info for sure.

**Reply by peterantypas** [reply]
This is how devices are programmed during production. There is a test pad on BOOT0 though that the jig drives high while resetting the MCU, so it does not depend on existing FW in any way.

The fully assembled unit does not expose RESET or BOOT0 though, but it does have the "dfu" command which jumps into the bootloader.

**Comment by peterantypas** [comment]
The basic UART breakout is in fab at JLCPCB now. I expect it sometime by the middle of next week. The firmware work should take me a day at most.

**Comment by Peter-Petrik** [comment]
I’m already running OpenPlotter on a RPi4 with PICAN-M HAT and one of the first MAIANA units with NMEA 2000. Will this new integrated solution work with that HAT and/or have other specifications, benefits worth an upgrade for my system?

**Reply by peterantypas** [reply]
I believe (and @sailoog can confirm) that for existing MAIANA installs all that will be needed is USB to the RPi and a firmware update for MAIANA itself. The idea is that Open Plotter will manage it via its serial interface.

**Reply by sailoog** [reply]
Correct. 

If you have a PICAN-M HAT or any in/out CAN converter connected to your Raspberry/OpenPlotter system, you do not need the MAIANA NMEA 2000 breakout, you just need the USB breakout to configure the transponder and get NMEA 0183 data from it. OpenPlotter and Signal K server will convert NMEA 0183 data into 2000 and will feed your NMEA 2000 network through your CAN converter.

In your case I guess your NMEA 2000 breakout is feeding your NMEA 2000 network and the PICAN-M HAT is getting data from there, so you can use the USB connection only for eventually configuring the MAIANA transponder.

**Comment by sailoog** [comment]
The main goal of this OP integration is to facilitate the configuration of the station data (even at runtime) and enabling/disabling transmission by software at anytime. We still have to confirm if we can also make updating the firmware easier.

**Comment by peterantypas** [comment]
First prototype of the bare-bones UART breakout is here. This one is meant for wiring to a Raspberry Pi, but it can also go on a breadboard, a protoboard, or be screwed onto a panel with the pins mounted upward. I think I'm going to change the holes for M2 mounts instead of M3, that was probably overkill :-)

![IMG_1329](https://user-images.githubusercontent.com/1565933/136293841-4ad1255a-6839-4631-8dbc-3901ce057379.JPG)

![IMG_1330](https://user-images.githubusercontent.com/1565933/136293865-260b6d14-1c81-49cb-8bdd-fe43eaf94ade.JPG)

**Comment by sailoog** [comment]
Looks great!
Size?
Could the screw connector eventually be replaced by 2 pins to be powered from a host board?

**Reply by peterantypas** [reply]
Sure, I could do that but do you think there will always be a "host" board?

**Reply by sailoog** [reply]
Sorry my question is ambiguous, I mean:
Do you think I could solder 2 pins instead that screw connector for my project involving a host board with a 12V rail?

**Comment by peterantypas** [comment]
Yes, of course. It’s a 0.1” pitch

**Comment by sailoog** [comment]
Perfect. thanks.

Sizes?

**Reply by peterantypas** [reply]
LOA: 60mm, beam: 18mm

**Reply by peterantypas** [reply]
height: 14mm

**Comment by sailoog** [comment]
A MAIANA UART kit just landed in Barcelona I will start with the integration this weekend :)

**Comment by sailoog** [comment]
Here we go...

This is my setup in the office:

![photo1636837757](https://user-images.githubusercontent.com/8145560/141659264-cd5f6756-e3b5-4953-b73c-89e6f3f7be1b.jpeg)
![photo1636837757(1)](https://user-images.githubusercontent.com/8145560/141659277-e5e180c6-f9e6-4989-b7b3-b4b3904929c6.jpeg)

On the left is the power source, in the middle is the MAIANA UART version (green led is AIS reception, blue led is GPS fix and red led is AIS transmission) and on the right is the Raspberry running OpenPlotter and getting data by UART

In OpenPlotter you have to make a serial connection in Signal K using the OpenPlotter serial app as usual. I have enabled the UART GPIO pins but if you have the MAIANA USB version you do not need to do this:

![Screenshot from 2021-11-13 21-03-41](https://user-images.githubusercontent.com/8145560/141659396-8053bbbd-2957-47ed-95e1-9115d7223337.png)

Finally you just have to select this Signal K connection in the new openPlotter MAIANA app and you will be able to communicate with your MAIANA device easily:

![Screenshot from 2021-11-13 20-59-59](https://user-images.githubusercontent.com/8145560/141659429-eaab47a5-9405-4c26-98ae-38c11d96d6f4.png)

The "Update firmware" feature is still under construction but I think I will publish a beta soon so the lucky ones who already have a MAIANA unit can play with it.

I am really impressed by the quality of the reception. I am inside a building 4 miles from the sea and I get the GPS fix after seconds and many AIS targets from 16 miles, including the small buoys in the port of Barcelona!

**Comment by sailoog** [comment]
This is how MAIANA data is stored in Signal K:

```
{
  "hardwareRevision": {
    "meta": {},
    "value": "11.3.0",
    "$source": "OpenPlotter.maiana",
    "timestamp": "2021-11-13T21:47:34.899Z"
  },
  "firmwareRevision": {
    "meta": {},
    "value": "3.3.0",
    "$source": "OpenPlotter.maiana",
    "timestamp": "2021-11-13T21:47:34.899Z"
  },
  "serialNumber": {
    "meta": {},
    "value": "",
    "$source": "OpenPlotter.maiana",
    "timestamp": "2021-11-13T21:47:34.899Z"
  },
  "MCUtype": {
    "meta": {},
    "value": "",
    "$source": "OpenPlotter.maiana",
    "timestamp": "2021-11-13T21:47:34.899Z"
  },
  "station": {
    "MMSI": {
      "meta": {},
      "value": "987654321",
      "$source": "OpenPlotter.maiana",
      "timestamp": "2021-11-13T21:47:34.910Z"
    },
    "vesselName": {
      "meta": {},
      "value": "TEST",
      "$source": "OpenPlotter.maiana",
      "timestamp": "2021-11-13T21:47:34.910Z"
    },
    "callSign": {
      "meta": {},
      "value": "ABCD",
      "$source": "OpenPlotter.maiana",
      "timestamp": "2021-11-13T21:47:34.910Z"
    },
    "vesselType": {
      "meta": {},
      "value": 30,
      "$source": "OpenPlotter.maiana",
      "timestamp": "2021-11-13T21:47:34.910Z"
    },
    "LOA": {
      "meta": {
        "description": "AC equipment common qualities"
      },
      "value": 12,
      "$source": "OpenPlotter.maiana",
      "timestamp": "2021-11-13T21:47:34.910Z"
    },
    "beam": {
      "meta": {},
      "value": 4,
      "$source": "OpenPlotter.maiana",
      "timestamp": "2021-11-13T21:47:34.910Z"
    },
    "portOffset": {
      "meta": {},
      "value": 1,
      "$source": "OpenPlotter.maiana",
      "timestamp": "2021-11-13T21:47:34.910Z"
    },
    "bowOffset": {
      "meta": {},
      "value": 11,
      "$source": "OpenPlotter.maiana",
      "timestamp": "2021-11-13T21:47:34.910Z"
    }
  },
  "transmission": {
    "hardwarePresent": {
      "meta": {},
      "value": 1,
      "$source": "OpenPlotter.maiana",
      "timestamp": "2021-11-13T21:47:34.922Z"
    },
    "hardwareSwitch": {
      "meta": {},
      "value": 1,
      "$source": "OpenPlotter.maiana",
      "timestamp": "2021-11-13T21:47:34.922Z"
    },
    "softwareSwitch": {
      "meta": {},
      "value": 1,
      "$source": "OpenPlotter.maiana",
      "timestamp": "2021-11-13T21:47:34.922Z"
    },
    "stationData": {
      "meta": {},
      "value": 1,
      "$source": "OpenPlotter.maiana",
      "timestamp": "2021-11-13T21:47:34.922Z"
    },
    "status": {
      "meta": {},
      "value": 1,
      "$source": "OpenPlotter.maiana",
      "timestamp": "2021-11-13T21:47:34.922Z"
    }
  },
  "channelA": {
    "noiseFloor": {
      "meta": {},
      "value": 50,
      "$source": "OpenPlotter.maiana",
      "timestamp": "2021-11-13T21:47:31.939Z"
    },
    "transmittedMessageType": {
      "meta": {},
      "value": "24A",
      "$source": "OpenPlotter.maiana",
      "timestamp": "2021-11-13T21:47:40.451Z"
    }
  },
  "channelB": {
    "noiseFloor": {
      "meta": {},
      "value": 46,
      "$source": "OpenPlotter.maiana",
      "timestamp": "2021-11-13T21:47:31.939Z"
    }
  }
}
```

**Comment by sailoog** [comment]
I have published the first beta. I have the MAIANA UART version so it would be great if someone can test it with any of the other MAIANA versions by USB. Short manual:

- Update OpenPlotter Settings app to 2.9.1
- After updating you will see a new app "MAIANA AIS trabnsponder". Install it.
- Open the OpenPlotter Serial app and create an alias for your MAIANA device and an "auto" connection in Signal K for it. From this moment you should get GPS/AIS data in Signal K and OpenCPN.
- Open the OpenPlotter MAIANA AIS app and approve the connection of this app with signal k with read/write permissions.
- Finally select the signal k connection for your MAIANA device in OpenPlotter MAIANA AIS app. Press refresh to get data from the device and "save station data" to send data to the device.
- Please report any issue here: https://github.com/openplotter/openplotter-maiana

Enjoy

**Reply by muebau** [reply]
I tried the software and it looks good. The only thing I found is that it deletes all configuration on the device. I am unable to write this configuration again even with my terminal. Is there something I should worry about? Do I need to flash the device now to get a working device again? What does the software send over USB other than I would send with my terminal?

**Reply by peterantypas** [reply]
That's strange. Can you elaborate further? Are you saying that sending the "station" command from the terminal no longer has an effect?

**Reply by muebau** [reply]
yes neither "cli" nor "station ..." and the new "MAIANA AIS trabnsponder" app seems to be unable too. It looks like the TX is kind of ignored. Is there a chance that some input of the USB TTY can harm MAIANA?

**Reply by sailoog** [reply]
really strange. According to the issue you opened in openplotter-maiana app: https://github.com/openplotter/openplotter-maiana/issues/1

that happened after trying to set the TX switch. That action just sends the command "tx?" to the device.

**Reply by sailoog** [reply]
rectification, sends the command "tx on" or "tx off"

**Reply by muebau** [reply]
hmmm that is good to hear. The App was fine and as expected (all values) but after that "tx" all values became "0" and "". Is there something I could try to help to find the cause?

**Comment by peterantypas** [comment]
I need to clarify here that this may only work for those who have board version 11.3 and newer. There’s only a few of those in circulation.

**Reply by kagouraki** [reply]
Is this a firmware problem or hardware ? The compatibility i mean.

**Comment by peterantypas** [comment]
It’s both. Some of the functionality required a hardware revision

**Comment by Rudder2** [comment]
Starting to wish I waited and knew this integration was coming and I might
of bought a different model. Will the NMEA 2k model work with this
integration in any way? I Use OpenPlotter as my ships navigation system.

**Reply by muebau** [reply]
I get the same feeling

**Comment by peterantypas** [comment]
Yeah it does. Email me, we’ll work something out

**Comment by Techstyleuk** [comment]
This is Great!!  thanks for all of your efforts

Just to clarify, do the UART and the (newer) USB versions have the same functionality with Openplotter?

1. AIS Transponder
2. GPS Data
3. Openplotter managed station data
4. Openplotter software TX switch
5. Openplotter managed Firmware update 

I use Bluetooth for streaming audio sometimes, I am not sure if UART and BT can be used at the same time?   I thought I had read somewhere that they shared some of the same resources.

**Reply by kagouraki** [reply]
You can use one at the time. Uart or BL. Not both at the same time

**Reply by sailoog** [reply]
Yes, no difference between UART and USB.

Yes, Bluetooth and UART share the same interface in Raspberry Pi. OpenPlotter serial app takes care of this disabling or enabling Bluetooth. If you need to use Bluetooth and a MAIANA device at the same time just use MAIANA NMEA 2000, NMEA 0183 or USB version, not the UART version.

**Comment by peterantypas** [comment]
No, I don't think so, something else is happening. Are you trying to use it
with a Pi, or a laptop?

**Reply by muebau** [reply]
The new "MAIANA AIS trabnsponder" app is on OpenPlotter on a RPi 4. My terminal is "GTKTerm" on my Ubuntu laptop. I used "screen /dev/ttyUSB0 38400" too.

**Reply by muebau** [reply]
just to be clear:
- power on the board ("$PAISYS,,*01 ....")
- type "cli" <enter> (CR LF) (nothing changes)
- type "station ........."  <enter> (CR LF) (nothing changes)
- no "$PAISTN," as reponse

Is there an obvious mistake?

**Comment by peterantypas** [comment]
I don't think the board he has implements this. It's version 10.9.0. It
should ignore it.

**Reply by muebau** [reply]
Where could I find the version?
The board greets me with 
```
$PAISYS,,*01
$PAISTN,0,,,0,0,0,0,0*11
$PMTK011,MTKGPS*08
$PMTK010,001*2E
$PMTK011,MTKGPS*08
$PMTK010,002*2D
$GPRMC,235942.293,V,,,,,0.00,0.00,050180,,,N*42
```
after power on.

**Comment by sailoog** [comment]
It would make sense that a non expected command could clear all station data, you never know what side effects you will have. But this permanent clear station data does not make sense...

**Comment by peterantypas** [comment]
Please DM me. This may be hardware damage.

**Comment by sailoog** [comment]
New version of openplotter-maiana app going from alpha to beta and probably a release candidate. Changes:

- Added electromagnetic interference detection (noise detection). It triggers Signal k notifications.
- Detect bad connections with MAIANA device
- Added firmware updating!
- Added minimum hardware and firmware versions detection
- Added translations to several languages

![Screenshot from 2021-11-25 20-12-13](https://user-images.githubusercontent.com/8145560/143496983-5ef8a72e-4181-4065-9e88-2fb81be1bdbc.png)

![Screenshot from 2021-11-25 20-13-10](https://user-images.githubusercontent.com/8145560/143496998-ba4d7416-27d3-41e6-ae9c-a636f00f8f3d.png)
